{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3eff5bcd-c545-4a20-b37b-30f0d56e1602",
   "metadata": {},
   "outputs": [],
   "source": [
    "!export CUBLAS_WORKSPACE_CONFIG=:4096:8\n",
    "!export CUDA_LAUNCH_BLOCKING=1   # (optional, helps debug determinism)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b99151ab-4029-40b5-8afd-9a74f5d95626",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HF_HOME: /projects/sciences/computing/sheju347/.cache/huggingface\n",
      "HF_HUB_CACHE: /projects/sciences/computing/sheju347/.cache/huggingface/hub\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Set env vars BEFORE importing huggingface modules\n",
    "os.environ[\"HF_HOME\"] = \"/projects/sciences/computing/sheju347/.cache/huggingface\"\n",
    "os.environ[\"TRANSFORMERS_CACHE\"] = \"/projects/sciences/computing/sheju347/.cache/huggingface/hub\"\n",
    "\n",
    "# Now import huggingface modules\n",
    "from huggingface_hub import constants\n",
    "\n",
    "print(\"HF_HOME:\", constants.HF_HOME)\n",
    "print(\"HF_HUB_CACHE:\", constants.HF_HUB_CACHE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "349a8d9c-2c15-4b55-815b-a696377a14ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/projects/sciences/computing/sheju347/miniconda3/envs/LLM311/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "/projects/sciences/computing/sheju347/miniconda3/envs/LLM311/lib/python3.11/site-packages/transformers/utils/hub.py:111: FutureWarning: Using `TRANSFORMERS_CACHE` is deprecated and will be removed in v5 of Transformers. Use `HF_HOME` instead.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from model_trainer import ModelTrainer\n",
    "from test_performance import TestPerformance, DatasetPath, MMLU_Subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "283a1066-2c5f-47e4-86ab-2b539229b3ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------- start checking GPU -----------\n",
      "GPU: NVIDIA H100 NVL\n",
      "torch.cuda.is_bf16_supported():  True\n",
      "---------- finish checking GPU -----------\n",
      "---------- start loading model:/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/base/10-15-UltraMedical-batchsize8-bf16 -----------\n",
      "finish loading tokenizer\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading checkpoint shards: 100%|██████████| 2/2 [00:00<00:00,  4.90it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "applying LoRA adapter from: /projects/sciences/computing/sheju347/MedicalQA/train/saved_models/LoRA/10-17-new-base-model-1e-4-batch8-LoRA-1e-4-r16\n",
      "LoRA adapter applied\n",
      "finish loading model\n",
      "torch_dtype: torch.bfloat16\n",
      "cuda available: True\n",
      "device: cuda\n",
      "---------- finish loading model:/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/base/10-15-UltraMedical-batchsize8-bf16 -----------\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "trainer = ModelTrainer()\n",
    "\n",
    "# model_name = \"microsoft/Phi-3-mini-4k-instruct\"\n",
    "# model_name = \"KrithikV/MedMobile\"\n",
    "# model_name = \"Qwen/Qwen3-4B-Instruct-2507\"\n",
    "# model_name = \"google/gemma-3-4b-it\"\n",
    "# model_name = \"meta-llama/Llama-3.2-3B-Instruct\"\n",
    "# model_name = \"google/medgemma-4b-it\"\n",
    "# model_name = \"Qwen/Qwen2.5-3B-Instruct\"\n",
    "# model_name = os.path.abspath(\"../../../projects/sciences/computing/sheju347/MedicalQA/train/saved_models/fine_tuned_model_entire_UltraMedical_batch_4\")\n",
    "# model_name = \"/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/fine_tuned_model_entire_UltraMedical_batch_4\"\n",
    "# model_name = \"/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/base/10-15-UltraMedical-batchsize8-lr2e-5-bf16\"\n",
    "model_name = \"/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/base/10-15-UltraMedical-batchsize8-bf16\" # BEST\n",
    "# model_name = \"/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/base/11-29-phi3-mini-UltraMedical-batchsize32\"\n",
    "# model_name = \"/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/base_qwen/11-11-Qwen3-4B-base-UltraMedical\"\n",
    "# model_name = \"/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/base_qwen/11-14-Qwen3-4B-base-UltraMedical\"\n",
    "# model_name = \"/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/fine_tuned_Phi_3_mini\"\n",
    "# model_name = \"/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/fine_tuned_model_no_mask_entire\"\n",
    "# model_name = \"/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/10-3-fine-tuned-all\"\n",
    "# model_name = \"/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/10-5-fine-tuned-UltraMedical-MedQA-context/\"\n",
    "# model_name = \"/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/10-6-fine-tuned-UltraMedical-MedQA-LoRA\"\n",
    "\n",
    "# lora_adapter_path = \"/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/10-6-fine-tuned-UltraMedical-MedQA-LoRA\"\n",
    "# lora_adapter_path = \"/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/10-10-LoRA-context012-r-16-lr-1e-5\"\n",
    "# lora_adapter_path = \"/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/10-12-LoRA-Evol-context012-r-16-lr-1e-5-epoch3/checkpoint-37846\"\n",
    "lora_adapter_path = \"/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/LoRA/10-17-new-base-model-1e-4-batch8-LoRA-1e-4-r16\" # BEST\n",
    "\n",
    "\n",
    "# model_name = \"google/flan-t5-xl\"\n",
    "\n",
    "# trainer.load_model(model_name)\n",
    "trainer.load_model(model_name, lora_adapter_path)\n",
    "# trainer.load_model(model_name, tokenizer_path = \"/projects/sciences/computing/sheju347/MedicalQA/train/saved_models/base_qwen/11-11-Qwen3-4B-base-UltraMedical\")\n",
    "\n",
    "# trainer.load_model_t5(model_name)\n",
    "# trainer.test_model_MedQA_response()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "da2f89ba-3ca3-4a20-a563-41e6aa1a6c3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "is_encoder_decoder = False\n",
    "\n",
    "test = TestPerformance(trainer.model, trainer.tokenizer, trainer.device, is_encoder_decoder = is_encoder_decoder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f50774f0-7b2f-47aa-9cec-7a32401c90c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "`trust_remote_code` is not supported anymore.\n",
      "Please check that the Hugging Face dataset 'openlifescienceai/medmcqa' isn't based on a loading script and remove `trust_remote_code`.\n",
      "If the dataset is based on a loading script, please ask the dataset author to remove it and convert it to a standard format like Parquet.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "use_RAG: True\n",
      "topK_searchEngine: 150\n",
      "topK_SPLADE: 30\n",
      "topK_denseEmbedding: 0\n",
      "topK_crossEncoder: 1\n",
      "topK_LLM: 0\n",
      "score_threshold: None\n",
      "RRF_models: None\n",
      "model: /projects/sciences/computing/sheju347/MedicalQA/train/saved_models/base/10-15-UltraMedical-batchsize8-bf16\n",
      "RAG data: [{'docId': 'pubmed23n0030_5432', 'BM25_score': 27.4436, 'BM25_ranking': 105, 'SPLADE_score': 21.397432327270508, 'SPLADE_ranking': 4, 'crossEncoder_score': np.float32(0.26192823), 'crossEncoder_ranking': 1}]\n",
      "outputs text: <|user|> Context:\n",
      "[Unmyelinated nerve fibers in senile nerves and in late thalidomide neuropathy: a quantitative electron microscopic study (author's transl)]. Sural nerve biopsies of four patients, aged 54--76 years, with a predominantly sensory type of neuropathy following high dosages of thalidomide were examined by light and electron microscopy. The present study includes a qualitative and quantitative evaluation of unmyelinated nerve fibers. Despite severe neuropathy, increased numbers of small unmyelinated axons per endoneurial area were noted in all patients. This numerical increase appeared to be independent of aging, since it was not seen in two senile controls, studied at the age of 83 and 88 years. The increase in the endoneurial density of unmyelinated axons, especially of small sized fibers, is likely to be related to regeneration following degeneration of unmyelinated axons although endoneurial shrinkage secondary to loss of large myelinated fibers could have caused an additional increase in the number of axons per endoneurial area. Axonal sprouting, despite degeneration of large numbers of myelinated and unmyelinated fibers, appears to be consistent with some of the characteristic clinical features of thalidomide neuropathy such as paresthesias, hyperesthesia for pain and temperature, and disturbances of autonomic functions. On the other hand, a variable number of empty Schwann cells (bands of BÃ¼ngner) and pockets at the surface of many Schwann cells noted in the four patients with neuropathy were also seen in both senile controls with no signs of neuropathy. Thus, it is obvious that pockets and empty Schwann cells may be related to aging or other causes of slow axonal wasting with Schwann cell proliferation and are not necessarily associated with clinically manifest neuropathy.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Question:\n",
      "Which of the following is not true for myelinated nerve fibers:\n",
      "\n",
      "[A] : Impulse through myelinated fibers is slower than non-myelinated fibers\n",
      "[B] : Membrane currents are generated at nodes of Ranvier\n",
      "[C] : Saltatory conduction of impulses is seen\n",
      "[D] : Local anesthesia is effective only when the nerve is not covered by myelin sheath\n",
      "<|end|><|assistant|> [A]. Impulse through myelinated fibers is slower than non-myelinated fibers - This statement is not true. Myelination actually increases the speed of impulse conduction along the nerve fiber. Myelin acts as an insulator and allows for faster transmission of electrical signals through a process called saltatory conduction, where the impulse jumps from one node of Ranvier to the next.\n",
      "\n",
      "[B]. Membrane currents are generated at nodes of Ranvier - This statement is true. The nodes of Ranvier are gaps in the myelin sheath where ion channels are concentrated. When an action potential reaches a node, it triggers a re-invocation of the action potential, allowing the nerve impulse to continue along the axon.\n",
      "\n",
      "[C]. Saltatory conduction of impulses is seen - This statement is true. As mentioned earlier, saltatory conduction is the process by which action potentials are propagated along myelinated axons in a jumping fashion from one node of Ranvier to the next.\n",
      "\n",
      "[D]. Local anesthesia is effective only when the nerve is not covered by myelin sheath - This statement is not true. Local anesthetics can be effective on myelinated nerves as well. They work by blocking sodium channels, which are present in both myelinated and non-myelinated nerve fibers. This prevents the propagation of action potentials along the nerve, leading to a loss of sensation in the area.\n",
      "\n",
      "So, the answer is [A].\n",
      "<|end|>\n",
      "question 1/4183 answer:A correct_answer:A True\n",
      "RAG data: [{'docId': 'pubmed23n0124_13424', 'BM25_score': 25.0574, 'BM25_ranking': 1, 'SPLADE_score': 20.217323303222656, 'SPLADE_ranking': 2, 'crossEncoder_score': np.float32(1.4490694), 'crossEncoder_ranking': 1}]\n",
      "outputs text: <|user|> Context:\n",
      "Theoretical effects of a distribution of capillary dimensions on glomerular ultrafiltration. Simple network models were developed to examine the effects of a distribution of capillary dimensions on the filtration of fluid in the renal glomerulus. Two idealized networks were considered, one with parallel capillaries of varying length, and the other with parallel capillaries of varying radius. It was found that for a fixed value of the ultrafiltration coefficient (Kf), the product of hydraulic permeability and total capillary surface area per glomerulus, any assumed heterogeneity of capillary dimensions lowered the single nephron glomerular filtration rate (SNGFR) below that predicted for uniform capillaries in parallel. This implies that, to the extent that the capillaries are not identical, a given filtering surface area is used less efficiently. A consequence of this is that for a given set of measured glomerular pressures and flows, a model which assumes identical capillaries will underestimate the value of Kf for the actual (heterogeneous) glomerular network. The effects of a distribution in capillary lengths were found to be more pronounced, in general, than those of a distribution in radii. Considering only heterogeneity in capillary length, and characterizing the effective length distribution using available data in the literature, it was found that calculating Kf using the customary assumption of identical capillaries will underestimate the true value by approximately 30%, under typical micropuncture conditions in the rat. The predicted dependence of SNGFR on afferent plasma flow rate, transmural hydraulic pressure difference, and afferent protein concentration was quite similar, however, whether capillary dimensions were assumed to be uniform or heterogeneous.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Question:\n",
      "Which of the following is not true about glomerular capillaries')\n",
      "\n",
      "[A] : The oncotic pressure of the fluid leaving the capillaries is less than that of fluid entering it\n",
      "[B] : Glucose concentration in the capillaries is the same as that in glomerular filtrate\n",
      "[C] : Constriction of afferent aeriole decreases the blood flow to the glomerulas\n",
      "[D] : Hematocrit of the fluid leaving the capillaries is less than that of the fluid entering it\n",
      "<|end|><|assistant|> [A]. The oncotic pressure of the fluid leaving the capillaries is less than that of fluid entering it - This is true because as fluid is filtered out of the capillaries into the Bowman's space, it leaves behind proteins and other solutes, which creates a lower oncotic pressure compared to the blood entering the glomerulus.\n",
      "\n",
      "[B]. Glucose concentration in the capillaries is the same as that in glomerular filtrate - This is not true. While glucose is freely filtered at the glomerulus, it is almost completely reabsorbed in the proximal tubule, so its concentration in the capillaries is not the same as in the glomerular filtrate.\n",
      "\n",
      "[C]. Constriction of afferent arteriole decreases the blood flow to the glomerulus - This is true. The afferent arteriole is the vessel that brings blood to the glomerulus. If it constricts, less blood will flow into the glomerulus, thereby decreasing the glomerular filtration rate.\n",
      "\n",
      "[D]. Hematocrit of the fluid leaving the capillaries is less than that of the fluid entering it - This is true because as blood is filtered, red blood cells are typically not filtered out, so the fluid leaving the glomerulus will have a lower concentration of red blood cells, or a lower hematocrit, compared to the blood entering the glomerulus.\n",
      "\n",
      "So, the answer is [B].\n",
      "<|end|>\n",
      "question 2/4183 answer:B correct_answer:A False\n",
      "RAG data: [{'docId': 'pubmed23n0207_857', 'BM25_score': 38.3574, 'BM25_ranking': 78, 'SPLADE_score': 55.841819763183594, 'SPLADE_ranking': 1, 'crossEncoder_score': np.float32(2.6872907), 'crossEncoder_ranking': 1}]\n",
      "outputs text: <|user|> Context:\n",
      "Reproduction in Down syndrome. A new case of pregnancy in a 29-year-old woman with trisomy 21 is described. She gave birth to a male infant, chromosomally and phenotypically normal, who died the day after delivery due to prematurity. Thirty pregnancies of 26 affected mothers, including the present woman, resulted in 10 children with Down syndrome, 18 children (1 set of twins) without Down syndrome, and 3 spontaneous abortions. Although rare, pregnancies in women with Down syndrome could become more frequent, increasing the importance of genetic, reproductive, and obstetric problems. These problems are reviewed with special reference to the occurrence of nonspecific abnormalities in chromosomally normal children, difficulties in labor and delivery, and frequency of prematurity and low birth weight.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Question:\n",
      "A 29 yrs old woman with a pregnancy of 17 week has a 10 years old boy with down syndrome. She does not want another down syndrome kid; best advice to her is\n",
      "\n",
      "[A] : No test is required now as her age is below 35 years\n",
      "[B] : Ultra sound at this point of time will definitely tell her that next baby will be down syndromic or not\n",
      "[C] : Amniotic fluid samples plus chromosomal analysis will definitely tell her that next baby will be down syndromic or not\n",
      "[D] : blood screening at this point of time will clear the exact picture\n",
      "<|end|><|assistant|> The woman in question is 29 years old and is currently pregnant with her second child. She has a 10-year-old son with Down syndrome, which is a genetic disorder caused by the presence of all or part of a third copy of chromosome 21.\n",
      "\n",
      "[A]. No test is required now as her age is below 35 years - This is incorrect because the woman's previous child has Down syndrome, which may indicate a genetic predisposition. Age is a risk factor for Down syndrome, but at 29, her age alone does not rule out the possibility of having another child with Down syndrome.\n",
      "\n",
      "[B]. Ultrasound at this point of time will definitely tell her that next baby will be down syndromic or not - This is incorrect because while certain ultrasound markers can indicate an increased risk of Down syndrome, ultrasound cannot definitively diagnose Down syndrome.\n",
      "\n",
      "[C]. Amniotic fluid samples plus chromosomal analysis will definitely tell her that next baby will be down syndromic or not - This is the most accurate option. Amniotic fluid sampling (also known as amniocentesis) can be performed around 15-20 weeks of pregnancy to obtain a sample of the amniotic fluid surrounding the fetus. This fluid contains fetal cells, which can be analyzed for chromosomal abnormalities, including Down syndrome. Chorionic villus sampling (CVS) is another invasive diagnostic test that can be done earlier, between 10-13 weeks of gestation, to obtain placental tissue for chromosomal analysis.\n",
      "\n",
      "[D]. Blood screening at this point in time will clear the exact picture - This is incorrect because blood screening tests, such as the first trimester combined test or the quadruple test, can only provide a risk assessment, not a definitive diagnosis. They are used as screening tools to determine the likelihood of the fetus having Down syndrome.\n",
      "\n",
      "So, the answer is [C].\n",
      "<|end|>\n",
      "question 3/4183 answer:C correct_answer:C True\n",
      "RAG data: [{'docId': 'pubmed23n0003_5314', 'BM25_score': 17.3698, 'BM25_ranking': 128, 'SPLADE_score': 27.596372604370117, 'SPLADE_ranking': 18, 'crossEncoder_score': np.float32(5.155279), 'crossEncoder_ranking': 1}]\n",
      "outputs text: <|user|> Context:\n",
      "Retrograde axonal and transsynaptic transport of macromolecules: physiological and pathophysiological importance. Anterograde and retrograde transport within axons and dendrites of nerve cells represent an integral part of the nerve cell function and biochemistry. A few exogenous macromolecules with most different molecular weights and physico-chemical properties (Nerve Growth Factor, tetanus toxin, cholera toxin, various lectins, antibodies against dopamine-beta-hydroxylase) have been shown to be taken up and transported with the retrograde axonal transport in exceedingly high amounts if compared to most other macromolecules. Specific binding to membrane receptors seems to be the prerequisite for this highly efficient retrograde transport. Upon arrival at the cell body tetanus toxin is able to leave the neuron and to migrate transsynaptically to presynaptic nerve terminals of second-order neurons. For NGF, tetanus toxin and some neurotropic viruses retrograde axonal transport eventually followed by transsynaptic transport may be crucially involved in their mechanism of action. Indirect evidence suggests the existence of a variety of endogenous molecules carrying specific information from the target cell and the nerve terminal to the cell body and eventually transsynaptically into second- or third-order neurons.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Question:\n",
      "Axonal transport is:\n",
      "\n",
      "[A] : Antegrade\n",
      "[B] : Retrograde\n",
      "[C] : Antegrade and retrograde\n",
      "[D] : None\n",
      "<|end|><|assistant|> Axonal transport is the process by which materials are moved along the axons of neurons. This transport can occur in both directions:\n",
      "\n",
      "- Antegrade transport refers to the movement of materials from the cell body (soma) to the axon terminals. This is typically used to transport newly synthesized proteins and organelles to the synapse.\n",
      "\n",
      "- Retrograde transport refers to the movement of materials from the axon terminals back to the cell body. This is often used for the recycling of materials and for signaling from the axon terminal to the cell body.\n",
      "\n",
      "Since axonal transport includes both antegrade and retrograde movement, the correct answer is that it encompasses both directions.\n",
      "\n",
      "So, the answer is [C].\n",
      "<|end|>\n",
      "question 4/4183 answer:C correct_answer:C True\n",
      "RAG data: [{'docId': 'pubmed23n0005_11882', 'BM25_score': 28.8901, 'BM25_ranking': 6, 'SPLADE_score': 23.75066375732422, 'SPLADE_ranking': 1, 'crossEncoder_score': np.float32(2.248272), 'crossEncoder_ranking': 1}]\n",
      "outputs text: <|user|> Context:\n",
      "Glucagon and insulin relationships in genetically diabetic (db/db) and in streptozotocin-induced diabetic mice. An inappropriate molar ratio of circulating insulin to glucagon is frequently associated with the metabolic alterations accompanying diabetes mellitus. Plasma immunoreactive insulin (IRI) and immunoreactive glucagon (IRG) levels were determined and the IRG:IRI ratio calculated at various intervals in overt diabetes in genetically diabetic (db/db) and in streptozotocin-treated mice. Plasma IRI levels in genetic mutants are elevated at nine weeks of age, but are comparable to values found in lean littermates by 21 weeks. The presence of a prevailing hyperglucagonemia is established for the first time in the intact db/db mice. Streptozotocin diabetics are found to have characteristically low plasma IRI and high plasma IRG values. The hormonal imbalance present in these two experimental animal models is accentuated when the data are expressed as the IRG:IRI ratio, which is seen to increase with the progression of diabetes.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Question:\n",
      "Low insulin to glucagon ratio is seen in all of these except:\n",
      "\n",
      "[A] : Glycogen synthesis\n",
      "[B] : Glycogen breakdown\n",
      "[C] : Gluconeogenesis\n",
      "[D] : Ketogenesis\n",
      "<|end|><|assistant|> Insulin and glucagon are two hormones that play a critical role in the regulation of glucose metabolism. Insulin is an anabolic hormone that promotes the storage of glucose as glycogen (glycogen synthesis), while glucagon is a catabolic hormone that stimulates the breakdown of glycogen into glucose (glycogen breakdown) and the production of glucose from non-carbohydrate sources (gluconeogenesis).\n",
      "\n",
      "A low insulin to glucagon ratio typically indicates a state where the body is breaking down glycogen (glycogen breakdown) or producing glucose from non-carbohydrate sources (gluconeogenesis). This ratio is not directly related to ketogenesis, which is the production of ketone bodies from fatty acids, primarily occurring when carbohydrate availability is low and insulin levels are low.\n",
      "\n",
      "Given the options:\n",
      "\n",
      "[A]. Glycogen synthesis - This process is stimulated by insulin, so a high insulin to glucagon ratio would be expected.\n",
      "[B]. Glycogen breakdown - This process is stimulated by glucagon, so a low insulin to glucagon ratio would be expected.\n",
      "[C]. Gluconeogenesis - This process is also stimulated by glucagon, so a low insulin to glucagon ratio would be expected.\n",
      "[D]. Ketogenesis - This process is stimulated by low insulin levels, but it is not directly related to the insulin to glucagon ratio in the context of glycogen metabolism.\n",
      "\n",
      "Since the question asks for the process that is not associated with a low insulin to glucagon ratio, the answer is the one that is not directly influenced by this ratio in the context of glycogen metabolism.\n",
      "\n",
      "So, the answer is [A].\n",
      "<|end|>\n",
      "question 5/4183 answer:A correct_answer:A True\n",
      "RAG data: [{'docId': 'pubmed23n0094_2394', 'BM25_score': 23.2324, 'BM25_ranking': 3, 'SPLADE_score': 31.065078735351562, 'SPLADE_ranking': 1, 'crossEncoder_score': np.float32(5.3478584), 'crossEncoder_ranking': 1}]\n",
      "outputs text: <|user|> Context:\n",
      "Ideal concentration of tropicamide with hydroxyamphetamine 1% for routine pupillary dilation. In this double-masked clinical study, we evaluated four concentrations of tropicamide (0.05%, 0.1%, 0.25%, and 0.5%) combined with hydroxyamphetamine 1% to find the combination that gives maximal pupillary dilation and inhibition of responsiveness to light and minimal paralysis of accommodation. With all concentrations, pupil size was maximal at 60 minutes, and there was no significant difference between the groups in mean pupillary diameter. Inhibition of the pupillary responses to light and loss of accommodation were directly related to the concentration of tropicamide. Tropicamide 0.25% combined with hydroxyamphetamine 1% was considered ideal for dilation and inhibition of the light response without inhibiting accommodation for near vision.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Question:\n",
      "Concentration of tropicamide:\n",
      "\n",
      "[A] : 0.01\n",
      "[B] : 0.02\n",
      "[C] : 0.03\n",
      "[D] : 0.04\n",
      "<|end|><|assistant|> Tropicamide is a muscarinic antagonist used in ophthalmology to dilate the pupil for examination of the retina. The concentration of tropicamide used for this purpose is typically 0.1%.\n",
      "\n",
      "Looking at the options provided:\n",
      "\n",
      "[A]. 0.01%\n",
      "[B]. 0.02%\n",
      "[C]. 0.03%\n",
      "[D]. 0.04%\n",
      "\n",
      "The correct concentration that matches the typical usage of 0.1% is not listed. However, the closest option to the actual concentration is 0.02%.\n",
      "\n",
      "So, the answer is [B].\n",
      "<|end|>\n",
      "question 6/4183 answer:B correct_answer:A False\n",
      "RAG data: [{'docId': 'pubmed23n0639_21146', 'BM25_score': 19.4535, 'BM25_ranking': 88, 'SPLADE_score': 14.326247215270996, 'SPLADE_ranking': 26, 'crossEncoder_score': np.float32(-3.3235781), 'crossEncoder_ranking': 1}]\n",
      "outputs text: <|user|> Context:\n",
      "Detection of influenza A H1N1 and H3N2 mutations conferring resistance to oseltamivir using rolling circle amplification. In the event of an influenza pandemic, the use of oseltamivir (OTV) will undoubtedly increase and therefore it is more likely that OTV-resistant influenza strains will also arise. OTV-resistance genotyping using sequence-based testing on viruses isolated in cell culture is time consuming and less likely to detect the low-level presence of drug-resistant virus populations. We have developed a novel rolling circle amplification (RCA) method to achieve the sensitive detection of OTV-resistant viruses from clinical specimens. Using artificially created templates, RCA could detect the presence of OTV-resistant mutations (N2: 119V, 292K, N1: 274Y) even if the population carrying the mutations was &lt;1% of the total. By applying RCA to clinical samples, we identified the emergence of the 274Y mutation in one OTV-treated patient, as well as in seven individuals who were treatment-naÃ¯ve (confirming community transmission of 274Y-containing resistant influenza A H1N1). These results were further confirmed by neuraminidase region sequencing. In conclusion, RCA technology can provide rapid (&lt;24 h), high-throughput diagnosis of OTV resistance mutations with a high specificity and sensitivity.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Question:\n",
      "Which of the following statements is true regarding H I N1 Influenza?\n",
      "\n",
      "[A] : Pregnant woman with sore throat can be staed immediately on oseltamivir without diagnostic testing under category B\n",
      "[B] : People on long-term steroids cannot receive Os-eltam ivir\n",
      "[C] : Category B concerns with low risk cases\n",
      "[D] : Category B patients have to undergo immediate testing\n",
      "<|end|><|assistant|> To determine the correct answer, let's analyze each statement in the context of H1N1 influenza and the use of oseltamivir, a medication used to treat and prevent influenza:\n",
      "\n",
      "[A]. Pregnant women with a sore throat can be started immediately on oseltamivir without diagnostic testing under category B: This statement is true. Pregnant women are considered a high-risk group for influenza, and prompt treatment with antivirals like oseltamivir is recommended without waiting for diagnostic confirmation. Category B refers to a category of drug use in pregnancy where animal reproduction studies have failed to demonstrate a risk to the fetus and there are no adequate and well-controlled studies in pregnant women, but the potential benefits may warrant use of the drug in pregnant women despite potential risks.\n",
      "\n",
      "[B]. People on long-term steroids cannot receive Oseltamivir: This statement is false. People on long-term steroids are not prohibited from receiving oseltamivir. However, they should be monitored closely because steroids can suppress the immune system, and the risk of influenza can be higher in these individuals.\n",
      "\n",
      "[C]. Category B concerns with low risk cases: This statement is misleading. Category B in the context of drug use during pregnancy indicates that animal reproduction studies have failed to demonstrate a risk to the fetus and there are no adequate and well-controlled studies in pregnant women, but the potential benefits may warrant use of the drug in pregnant women despite potential risks. It does not specifically concern low-risk cases.\n",
      "\n",
      "[D]. Category B patients have to undergo immediate testing: This statement is not necessarily true. While immediate testing may be warranted in some cases, the decision to test for H1N1 influenza or start oseltamivir should be based on clinical judgment and the patient's risk factors, not solely on being a Category B patient.\n",
      "\n",
      "Given the context of the question and the need to select the correct answer option A, we can conclude that the statement \"Pregnant woman with sore throat can be started immediately on oseltamivir without diagnostic testing under category B\" is true regarding H1N1 influenza.\n",
      "\n",
      "So, the answer is [A].\n",
      "<|end|>\n",
      "question 7/4183 answer:A correct_answer:A True\n",
      "RAG data: [{'docId': 'pubmed23n0600_899', 'BM25_score': 39.4, 'BM25_ranking': 1, 'SPLADE_score': 16.20403480529785, 'SPLADE_ranking': 3, 'crossEncoder_score': np.float32(2.609311), 'crossEncoder_ranking': 1}]\n",
      "outputs text: <|user|> Context:\n",
      "Nerve communication between the glossopharyngeal nerve, external carotid plexus and the superficial cervical ansa: human autopsy case. The authors encountered a very rare human autopsy case in which the supernumerary branch of the glossopharyngeal nerve and a nerve branch arising from the external carotid plexus communicated with the superficial cervical ansa. This anomaly was observed on the left side of a 71-year-old male cadaver during the gross anatomical seminar at Niigata University in 2004. The nerve fascicle and fiber analyses indicated that the supernumerary branch of the glossopharyngeal nerve separated cranial to the branches to the pharyngeal constrictor muscles, carotid sinus and stylopharyngeal muscle and sent the nerve fibers to the muscular branches to the platysma and the cutaneous branches to the cervical region. Additionally, it was shown that the branch arising from the external carotid plexus sent the nerve fibers to the cutaneous branch to the cervical region. Although the external carotid plexus is primarily postganglionic sympathetic fibers originating from the superior cervical ganglion, the vagus and glossopharyngeal nerves gave off branches connecting to the plexus, and therefore it was not possible to determine the origins of this branch of the external carotid plexus. The present nerve fascicle analysis demonstrates that the supernumerary branch of the glossopharyngeal nerve, which innervated the platysma, did not share any nerve components with the branches to the pharyngeal constrictor muscles, carotid sinus and stylopharyngeal muscle, suggesting that this supernumerary branch may be categorized into the different group from these well-known branches.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Question:\n",
      "Which of the following are not a branch of external carotid Aery in Kiesselbach's plexus.\n",
      "\n",
      "[A] : Sphenopalatine aery\n",
      "[B] : Anterior ethmoidal aery\n",
      "[C] : Greater palatine aery\n",
      "[D] : Septal branch of superior labial aery\n",
      "<|end|><|assistant|> Kiesselbach's plexus, also known as Little's area, is a region in the anterior part of the nasal septum where several arteries anastomose. This area is prone to nosebleeds (epistaxis) because of the rich vascular supply and the exposure of these vessels when the nasal mucosa is dry or irritated.\n",
      "\n",
      "The arteries that contribute to Kiesselbach's plexus include:\n",
      "\n",
      "[A]. Sphenopalatine artery - This is a branch of the maxillary artery, which is a branch of the external carotid artery. It does contribute to the blood supply of Kiesselbach's plexus.\n",
      "\n",
      "[B]. Anterior ethmoidal artery - This artery is actually a branch of the ophthalmic artery, which comes from the internal carotid artery, not the external carotid artery. Therefore, it does not contribute to Kiesselbach's plexus.\n",
      "\n",
      "[C]. Greater palatine artery - This artery is a branch of the maxillary artery, which is a branch of the external carotid artery. It does contribute to the blood supply of Kiesselbach's plexus.\n",
      "\n",
      "[D]. Septal branch of the superior labial artery - The superior labial artery is a branch of the facial artery, which is a branch of the external carotid artery. It does contribute to the blood supply of Kiesselbach's plexus.\n",
      "\n",
      "So, the answer is [B].\n",
      "<|end|>\n",
      "question 8/4183 answer:B correct_answer:B True\n",
      "RAG data: [{'docId': 'pubmed23n0510_16665', 'BM25_score': 16.5898, 'BM25_ranking': 41, 'SPLADE_score': 23.696552276611328, 'SPLADE_ranking': 2, 'crossEncoder_score': np.float32(1.5524276), 'crossEncoder_ranking': 1}]\n",
      "outputs text: <|user|> Context:\n",
      "A case of non-conducted atrial bigeminy simulating a second-degree atrioventricular block. A holter ECG diagnosis. A 82-years-old man, symptomatic for fatigue and lypothymia, was referred to our centre in order to evaluate the opportunity of a permanent pacemaker insertion. A 12-lead ECG was described as a Mobitz type II second-degree atrioventricular block with 2:1 conduction. This surface ECG revealed some sinus beats with normal AV conduction, everyone followed by a bizarre, non-conducted P' wave, inscripted in the previous T wave. The Holter ECG showed a sinus rhythm with a mean rate of 70 beats/minute: during the night and at 06:39 pm of the following day, ECG strip showed some sinus beats, everyone followed by an atrial non-conducted ectopic beat, characterized by prematurity and abnormal shape (P' wave), that appears as a small deformation on the preceding T wave. At the beginning and the end of the strip we can estimate respectively 9 and 4 sinus beats, that represent the normal rhythm of this patient. So, it is possible to compare the normal P-P interval (P-P = 0.84 sec) to the shorter P-P' interval (P-P = 0.40 sec) and make the correct diagnosis of non-conducted atrial bigeminy simulating a second-degree AV block with 2:1 conduction.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Question:\n",
      "Diagnosis of the following ECG-\n",
      "\n",
      "[A] : Ventricular bigeminy\n",
      "[B] : Electrical alternans\n",
      "[C] : P pulmonale\n",
      "[D] : Left ventricular failure\n",
      "<|end|><|assistant|> To diagnose the condition based on an ECG, we need to understand what each of the terms in the options refers to:\n",
      "\n",
      "[A]. Ventricular bigeminy - This is a pattern where every other heartbeat is a premature ventricular contraction (PVC). It is characterized by a normal beat followed by a premature ventricular contraction.\n",
      "\n",
      "[B]. Electrical alternans - This is a pattern seen on the ECG where there is a regular alternation in the amplitude or height of the QRS complexes. It is often associated with conditions like pericardial effusion.\n",
      "\n",
      "[C]. P pulmonale - This term refers to the P wave in lead P of the ECG that is greater than 2.5 mm in amplitude. It suggests right atrial enlargement, often due to pulmonary conditions.\n",
      "\n",
      "[D]. Left ventricular failure - This condition can lead to various ECG changes, but the classic ECG finding is not a specific pattern. It may show signs of strain or low voltage, but it is not diagnosed solely on the basis of ECG.\n",
      "\n",
      "Without the actual ECG image or description, we cannot definitively diagnose the condition. However, based on the typical ECG findings associated with each condition, we can infer the most likely answer.\n",
      "\n",
      "If the ECG showed a pattern of normal beats followed by premature ventricular contractions, the diagnosis would be ventricular bigeminy.\n",
      "\n",
      "So, the answer is [A].\n",
      "<|end|>\n",
      "question 9/4183 answer:A correct_answer:B False\n",
      "RAG data: [{'docId': 'pubmed23n0809_19167', 'BM25_score': 58.0771, 'BM25_ranking': 1, 'SPLADE_score': 59.441287994384766, 'SPLADE_ranking': 9, 'crossEncoder_score': np.float32(1.3918796), 'crossEncoder_ranking': 1}]\n",
      "outputs text: <|user|> Context:\n",
      "A rare case of congenital complex pulmonary AV fistula. A 15-year-old boy presented with central cyanosis with clubbing and dyspnoea on exertion. Cardiovascular examination did not reveal any abnormality. ECG was normal. Chest X-ray showed a normal sized heart with rounded opacities of variable size in the left upper lung field. Two-dimensional echocardiographic examination was normal. CT angiography showed a large complex lesion composed of serpiginous tubular structures involving the left upper and lingular lobe, suggestive of racemose tangle of blood vessels. A diagnosis of large complex arteriovenous (AV) fistula involving the left upper and lingular lobe was performed. This case reports a rare case of complex pulmonary AV fistula.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Question:\n",
      "A blue new born presents with cyanosis. The X–ray chest reveal oligaemic lung field and normal sized heart. Most likely diagnosis is –\n",
      "\n",
      "[A] : Ebstein's anomaly\n",
      "[B] : Pulmonary atresia\n",
      "[C] : Transposition of great arteries\n",
      "[D] : Tetralogy of fallot\n",
      "<|end|><|assistant|> A blue newborn with cyanosis and an X-ray showing oligemic lung fields with a normal-sized heart suggests a congenital heart defect that is causing poor oxygenation of the blood.\n",
      "\n",
      "[A]. Ebstein's anomaly is a rare heart defect in which the tricuspid valve is abnormally formed. This condition can lead to heart enlargement and heart failure, but it does not typically present with oligemic lung fields on an X-ray.\n",
      "\n",
      "[B]. Pulmonary atresia is a condition where the pulmonary valve is completely closed or absent, preventing blood from flowing from the right ventricle to the lungs. This would lead to cyanosis and oligemic lung fields due to the lack of blood flow to the lungs.\n",
      "\n",
      "[C]. Transposition of the great arteries is a condition where the two main arteries leaving the heart are switched in position. This results in oxygen-poor blood being circulated to the body and oxygen-rich blood being circulated back to the lungs. This condition can present with cyanosis and oligemic lung fields because the lungs are not receiving the blood they need to be aerated properly.\n",
      "\n",
      "[D]. Tetralogy of Fallot is a congenital heart defect that includes four abnormalities: a ventricular septal defect, pulmonary stenosis, right ventricular hypertrophy, and an overriding aorta. It can cause cyanosis and may show oligemic lung fields due to decreased blood flow to the lungs.\n",
      "\n",
      "Given the information provided, the most likely diagnosis is Transposition of the great arteries (C), as it is the condition that most directly explains the combination of cyanosis and oligemic lung fields with a normal-sized heart on an X-ray.\n",
      "\n",
      "So, the answer is [C].\n",
      "<|end|>\n",
      "question 10/4183 answer:C correct_answer:B False\n",
      "RAG data: [{'docId': 'pubmed23n0020_6161', 'BM25_score': 25.3413, 'BM25_ranking': 28, 'SPLADE_score': 17.76125717163086, 'SPLADE_ranking': 1, 'crossEncoder_score': np.float32(0.242535), 'crossEncoder_ranking': 1}]\n",
      "outputs text: <|user|> Context:\n",
      "Stages in the dissolution of human enamel crystals in dental caries. Various stages in monocrystal dissolution occurring during human enamel were studied by high resolution transmission electron microscopy. After the development of a central core lesion, two mechanisms by which the dissolution spread laterally to the (100) faces of the crystal could be demonstrated on the basis of the systematic orientation of the crystallographic axes. In the first case, the destruction was developing parallel to (120) planes and the borders were limited by (100) planes. In the second type, the development of the lateral side lesion was observed parallel to (100) planes. The carious destruction of the enamel monocrystal occurred as a result of the development of several lateral side dissolutions of the two types described, proceeding along the entire central core lesion.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Question:\n",
      "27.\tThe lateral spread of dental caries is facilitated mostly  by the\n",
      "\n",
      "[A] : Enamel spindles\n",
      "[B] : Dentinoenamel junction\n",
      "[C] : Enamel lamellae\n",
      "[D] : Striae of Retzius\n",
      "<|end|><|assistant|> Dental caries, commonly known as tooth decay, is a process where the hard tissues of the teeth are broken down by the acid produced by bacteria. The progression of dental caries can be influenced by various factors, including the structure and composition of the tooth enamel and dentin.\n",
      "\n",
      "[A]. Enamel spindles are the odontoblastic processes that extend into the enamel prism interstices. They can be a pathway for bacteria to reach the dentin, but they are not primarily responsible for the lateral spread of caries.\n",
      "\n",
      "[B]. The dentinoenamel junction (DEJ) is the interface between the enamel and dentin. While it is an important structural feature of the tooth, it is not the main facilitator of the lateral spread of caries.\n",
      "\n",
      "[C]. Enamel lamellae are microfissures within the enamel that can act as weak points where caries can initiate and spread. However, they are not the primary facilitators of the lateral spread of caries.\n",
      "\n",
      "[D]. Striae of Retzius are incremental lines in the enamel that represent periods of enamel formation. They can serve as pathways for caries to spread laterally because the enamel does not remodel after these lines, making them potential weak points for caries to exploit.\n",
      "\n",
      "Considering the options, the Striae of Retzius are the most likely to facilitate the lateral spread of dental caries because they represent areas where the enamel may be more susceptible to breakdown and where bacteria can invade and spread along these lines.\n",
      "\n",
      "So, the answer is [D].\n",
      "<|end|>\n",
      "question 11/4183 answer:D correct_answer:B False\n",
      "RAG data: [{'docId': 'pubmed23n0477_13094', 'BM25_score': 46.4471, 'BM25_ranking': 2, 'SPLADE_score': 30.673173904418945, 'SPLADE_ranking': 3, 'crossEncoder_score': np.float32(-1.4772997), 'crossEncoder_ranking': 1}]\n",
      "outputs text: <|user|> Context:\n",
      "The effect of arterial blood gas values on extubation decisions. No studies have examined the clinical utility of arterial blood gas (ABG) values during spontaneous breathing trials (SBTs) for making extubation decisions. Nonetheless many intensive care units measure ABGs during an SBT to determine, in conjunction with other data, whether the SBT was successful. Determine how often ABG values change extubation decisions. Each patient was assessed at the bedside by one of 3 critical care attending physicians. Each assessment was made at the end of the SBT in which ABGs were obtained with the expectation of extubation. When a nurse, respiratory therapist, or resident was available, he or she also simultaneously and independently evaluated the patient and then reassessed with the ABG values. Physiologic data were recorded before and during the SBT, as the attending physician performed the bedside assessment. Eighty-three patients were studied during 100 SBTs. Heart rate, respiratory rate, and mean blood pressure increased. Blood oxygen saturation (measured via pulse oximetry [F(IO)(2)]) and the P(aO)(2)/F(IO)(2) ratio decreased significantly during SBTs. Concordant clinical and ABG assessment (to attempt a trial of extubation) led to extubation in 86 patients, 76 of whom remained extubated after 72 hours. Following 14 of the SBTs, the patients were not extubated. Seven of these were judged by the attending physicians to fail both clinically and following review of ABG values. Seven were judged successful clinically, but the patients were not extubated because the attending physician thought the ABG values were unfavorable. Bedside nurses, who assessed 72 SBTs, had 1 case in which the ABG values changed their extubation opinion. Resident-trainees assessed 76 cases and had 5 changes of opinion based on ABG values. Respiratory therapists assessed 77 patients and had 4 such cases. The average changes in heart rate and blood pressure of patients not extubated were double those of extubated patients, but there were no significant differences in respiratory rate or P(aO)(2)/F(IO)(2) ratio between those groups. Ten patients (11.6%) were extubated based on the attending physician's decision but required re-intubation within 72 hours. These data suggest that ABG values did not change extubation decisions in 93% of cases. However, in 7 cases the ABG values changed the extubation decision. If even a few of those cases would have failed extubation without knowledge of the ABG values, the increased patient risk and cost associated with failed extubation would more than offset the relatively small cost of collecting ABG values from all patients who undergo SBT.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Question:\n",
      "A second-year PG resident tells you to perform an ABG of a patient. All of the following are true about performing an ABG except:\n",
      "\n",
      "[A] : Before performing the ABG, syringe should be loaded with 0.3 cc of heparin\n",
      "[B] : Normal pH, HCO. and PCO, levels may not indicate absence of an acid-base imbalance\n",
      "[C] : A different site should be tried i f modified Allen's test is negative\n",
      "[D] : Radial aery is the preferred site\n",
      "<|end|><|assistant|> [A]. Before performing the ABG, a syringe should be loaded with 0.3 cc of heparin - This is true. Heparin is added to the syringe to prevent clotting of the blood sample before analysis.\n",
      "\n",
      "[B]. Normal pH, HCO3-, and PCO2 levels may not indicate the absence of an acid-base imbalance - This is also true. The body has compensatory mechanisms that can maintain normal pH even in the presence of an underlying acid-base disorder. Therefore, normal ABG values do not always mean there is no imbalance.\n",
      "\n",
      "[C]. A different site should be tried if a modified Allen's test is negative - This statement is not true. The Allen's test is used to evaluate the adequacy of collateral blood flow to the hand before performing arterial blood gas sampling. A negative Allen's test indicates insufficient collateral flow, and the arterial puncture should be avoided to prevent ischemic complications. If the Allen's test is negative, the radial artery is not the preferred site, and the procedure should not be performed.\n",
      "\n",
      "[D]. Radial artery is the preferred site - This is true. The radial artery is the preferred site for ABG because it is easily accessible and has collateral blood flow, reducing the risk of ischemic complications.\n",
      "\n",
      "So, the answer is [C].\n",
      "<|end|>\n",
      "question 12/4183 answer:C correct_answer:A False\n",
      "RAG data: [{'docId': 'pubmed23n0561_10322', 'BM25_score': 26.8642, 'BM25_ranking': 9, 'SPLADE_score': 24.523059844970703, 'SPLADE_ranking': 1, 'crossEncoder_score': np.float32(5.939535), 'crossEncoder_ranking': 1}]\n",
      "outputs text: <|user|> Context:\n",
      "A simple model of dynamic interactions between respiratory centers. There is a general consensus that neurons responsible for the generation of a respiratory rhythm are located in the rostral ventrolateral medulla. However, there is still controversy regarding the necessary components for respiratory rhythm generation. Both inspiratory neurons in the preBotzinger Complex (preBOtC) and pre-inspiratory neurons located more rostrally to this anatomical structure referred to as the parafacial respiratory group (pFRG) have been proposed to be essential for respiratory rhythmogenesis. To study the dynamical interactions between preBOtC and pFRG neurons, we use a canonical model that describes each neuron population as a phase oscillator. We assume that the oscillators are weakly coupled with pFRG neurons providing stimulation to preBOtC neurons and preBOtC neurons providing inhibitory drive to pFRG neurons. In our mathematical study, we explore plausible mechanisms that may account for the complex interactions between I and pre-I neuron. In particular, we show that reduced excitability of inspiratory in preBOtC may lead to the phenomena known as \"quantal slowing\".\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Question:\n",
      "Respiratory rhythm generation center is located at:\n",
      "\n",
      "[A] : Dorsal respiratory group\n",
      "[B] : Pre-Botzinger complex\n",
      "[C] : Ventral respiratory neurons\n",
      "[D] : Pneumotaxic center\n",
      "<|end|><|assistant|> The respiratory rhythm generation is a complex process that involves various parts of the brainstem. The primary center responsible for generating the basic rhythm of respiration is the Pre-Bötzinger complex. This area is located in the ventrolateral medulla and has been shown to be critical for the generation of the respiratory rhythm in mammals.\n",
      "\n",
      "The dorsal respiratory group (A) is involved in the processing of sensory information and the modulation of the respiratory rhythm, but it is not the primary generator of the rhythm.\n",
      "\n",
      "The ventral respiratory neurons (C) are part of the network that helps control the muscles of respiration, but again, they are not the primary rhythm generators.\n",
      "\n",
      "The pneumotaxic center (D), located in the upper pons, helps regulate the rate and pattern of breathing but does not generate the respiratory rhythm itself.\n",
      "\n",
      "So, the answer is [B].\n",
      "<|end|>\n",
      "question 13/4183 answer:B correct_answer:B True\n",
      "RAG data: [{'docId': 'pubmed23n0571_23019', 'BM25_score': 31.5684, 'BM25_ranking': 1, 'SPLADE_score': 18.134069442749023, 'SPLADE_ranking': 1, 'crossEncoder_score': np.float32(0.853408), 'crossEncoder_ranking': 1}]\n"
     ]
    }
   ],
   "source": [
    "# data_range = [9, 13, 19, 24, 28, 29, 30, 31, 32, 38, 43, 44, 45, 46, 52, 56, 57, 62, 64, 68, 71, 73, 77, 87, 90, 94, 96, 97, 108, \n",
    "#               109, 110, 118, 123, 126, 133, 135, 139, 141, 143, 144, 145, 146, 159, 160, 166, 167, 170, 171, 172, 173, 179, 180, 181, 183, \n",
    "#               185, 186, 188, 191, 196, 197, 203, 205]\n",
    "\n",
    "# data_range = range(30, 32) # None #range(34, 35)\n",
    "\n",
    "data_range = None\n",
    "\n",
    "# data_range = range(7333, 10178)\n",
    "\n",
    "# data_range = range(1225, 1273)\n",
    "\n",
    "# data_range = range(2222, 4183)\n",
    "\n",
    "# data_range = range(298, 1000)\n",
    "\n",
    "topK_searchEngine = 150\n",
    "topK_SPLADE = 30\n",
    "topK_denseEmbedding = 0\n",
    "topK_crossEncoder = 1\n",
    "topK_LLM = 0\n",
    "score_threshold = None #0.7 # 0.8\n",
    "pick_rag_index = None\n",
    "get_classifier_training_data = False\n",
    "use_classifier = False\n",
    "RRF_models = None #[(\"SPLADE\", 0.8), (\"MonoT5\", 1), (\"Vicuna\", 1)] # [\"SPLADE\", \"DenseEmbedding\", \"MonoT5\", \"Zephyr\", \"Vicuna\"]\n",
    "\n",
    "# file_name = \"9-30-test_data_150_30_5_H100.txt\"\n",
    "# file_name = \"a.txt\"\n",
    "# file_name = \"10-14-lora-context012-r16-epoch3-150-30-5-all-data.txt\"\n",
    "# file_name = \"10-21-batchsize8-lr1e-4-bf16-new-lora-150-30-5-RRF.txt\"\n",
    "# file_name = \"11-4-150-30-5-1-MedQA-ensemble.txt\"\n",
    "# file_name = \"11-7-no-rag-ensemble-MedQA-A100.txt\"\n",
    "# file_name = \"11-13-Phi3-150-dense30-monoT51.txt\"\n",
    "# file_name = \"11-16-Phi3-1000-300-80-5-1.txt\"\n",
    "# file_name = \"11-17-Phi3-150-30-1-Vicuna-MMLU-Professional_medicine.txt\"\n",
    "# file_name = \"11-19-Phi3-150-30-1-Vicuna-MedMCQA.txt\"\n",
    "# file_name = \"11-20-llama3.2-3b-MedQA.txt\"\n",
    "# file_name = \"11-20-no-rag-College_medicine\"\n",
    "# file_name = \"11-21-Phi3-150-30-1-Dense-train.txt\"\n",
    "# file_name = \"11-25-Phi3-150-30-1-Vicuna-train.txt\"\n",
    "# file_name = \"11-25-medgemma-4b-MedQA.txt\"\n",
    "# file_name = \"11-25-flan-t5-xl-150-30-1.txt\"\n",
    "# file_name = \"11-26-Phi2.5-3b-no-rag-MedQA.txt\"\n",
    "# file_name = \"11-27-Phi3-150-1-bge-v2-m3.txt\"\n",
    "# file_name = \"12-1-Phi3-batchsize32-no-rag-MedQA.txt\"\n",
    "file_name = \"12-1-Phi3-150-30-1-MiniLM-MedMCQA.txt\"\n",
    "\n",
    "# test.SPLIT = \"train\"\n",
    "test.test_accuracy(DatasetPath.MedMCQA, subset_name = None, is_ensemble = False, use_RAG = True, data_range = data_range, file_name = file_name,\n",
    "                  topK_searchEngine = topK_searchEngine, topK_SPLADE = topK_SPLADE, topK_denseEmbedding = topK_denseEmbedding, topK_crossEncoder = topK_crossEncoder, topK_LLM = topK_LLM,\n",
    "                   score_threshold = score_threshold, pick_rag_index = pick_rag_index, get_classifier_training_data = get_classifier_training_data,\n",
    "                  use_classifier = use_classifier, RRF_models = RRF_models)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "421bb94f-c47d-4e92-9503-a873737c72fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data_range = [9, 13, 19, 24, 28, 29, 30, 31, 32, 38, 43, 44, 45, 46, 52, 56, 57, 62, 64, 68, 71, 73, 77, 87, 90, 94, 96, 97, 108, \n",
    "#               109, 110, 118, 123, 126, 133, 135, 139, 141, 143, 144, 145, 146, 159, 160, 166, 167, 170, 171, 172, 173, 179, 180, 181, 183, \n",
    "#               185, 186, 188, 191, 196, 197, 203, 205]\n",
    "\n",
    "# data_range = range(4, 5) # None #range(34, 35)\n",
    "\n",
    "data_range = None\n",
    "\n",
    "# data_range = range(763, 1273)\n",
    "\n",
    "topK_searchEngine = 150\n",
    "topK_SPLADE = 30\n",
    "topK_crossEncoder = 3\n",
    "topK_LLM = 0\n",
    "score_threshold = None # 0.8\n",
    "pick_rag_index = 2\n",
    "\n",
    "file_name = \"9-5-150_30_3rd_RAG_data_H100.txt\"\n",
    "# file_name = \"a.txt\"\n",
    "\n",
    "test.test_accuracy(DatasetPath.MedQA, subset_name = None, is_ensemble = False, use_RAG = True, data_range = data_range, file_name = file_name,\n",
    "                  topK_searchEngine = topK_searchEngine, topK_SPLADE = topK_SPLADE, topK_crossEncoder = topK_crossEncoder, topK_LLM = topK_LLM,\n",
    "                   score_threshold = score_threshold, pick_rag_index = pick_rag_index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c82f5ba-8f30-403f-bd01-4d8205710349",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data_range = [9, 13, 19, 24, 28, 29, 30, 31, 32, 38, 43, 44, 45, 46, 52, 56, 57, 62, 64, 68, 71, 73, 77, 87, 90, 94, 96, 97, 108, \n",
    "#               109, 110, 118, 123, 126, 133, 135, 139, 141, 143, 144, 145, 146, 159, 160, 166, 167, 170, 171, 172, 173, 179, 180, 181, 183, \n",
    "#               185, 186, 188, 191, 196, 197, 203, 205]\n",
    "\n",
    "# data_range = range(4, 5) # None #range(34, 35)\n",
    "\n",
    "data_range = None\n",
    "\n",
    "# data_range = range(763, 1273)\n",
    "\n",
    "topK_searchEngine = 150\n",
    "topK_SPLADE = 30\n",
    "topK_crossEncoder = 5\n",
    "topK_LLM = 0\n",
    "score_threshold = None # 0.8\n",
    "pick_rag_index = 4\n",
    "\n",
    "file_name = \"9-5-150_30_5th_RAG_data_H100.txt\"\n",
    "# file_name = \"a.txt\"\n",
    "\n",
    "test.test_accuracy(DatasetPath.MedQA, subset_name = None, is_ensemble = False, use_RAG = True, data_range = data_range, file_name = file_name,\n",
    "                  topK_searchEngine = topK_searchEngine, topK_SPLADE = topK_SPLADE, topK_crossEncoder = topK_crossEncoder, topK_LLM = topK_LLM,\n",
    "                   score_threshold = score_threshold, pick_rag_index = pick_rag_index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95867a73-0d71-4739-8d85-68ba0cd8b1ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "model_prompt = \"\"\"You are a medical question answering assistant.\n",
    "\n",
    "\n",
    "\n",
    "Question:\n",
    "A 2-year-old boy is brought to the physician because of decreased appetite and abdominal pain for the last several weeks. Physical examination shows a well-appearing toddler with a palpable left-sided abdominal mass that does not cross the midline. A CT of the abdomen shows a large, necrotic tumor on the left kidney. Histological examination of the kidney mass shows primitive blastemal cells and immature tubules and glomeruli. This tissue is most likely derived from the same embryological structure as which of the following?\n",
    "\n",
    "[A] : Adrenal medulla\n",
    "[B] : Thyroid gland\n",
    "[C] : Papillary muscles\n",
    "[D] : Anterior pituitary\"\"\"\n",
    "\n",
    "messages = [{\"role\": \"user\", \"content\": f\"{model_prompt}\"}]\n",
    "\n",
    "inputs = trainer.tokenizer.apply_chat_template(messages, add_generation_prompt = True, return_tensors = \"pt\").to(trainer.device)\n",
    "\n",
    "generate_kwargs = { \"max_new_tokens\": 1024, \"do_sample\": False }\n",
    "\n",
    "outputs = trainer.model.generate(inputs, **generate_kwargs)\n",
    "\n",
    "text = trainer.tokenizer.batch_decode(outputs)[0]\n",
    "\n",
    "print(text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5dc516c-a20d-4ff8-b66e-07af418003ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import ast\n",
    "import numpy as np\n",
    "\n",
    "QUESTION_COUNT = 1273\n",
    "\n",
    "path = \"\"\n",
    "file_names = [\"9-5-150_30_1_RAG_data_H100.txt\", \"9-5-150_30_2nd_RAG_data_H100.txt\", \"9-5-150_30_3rd_RAG_data_H100.txt\", \n",
    "              \"9-5-150_30_4th_RAG_data_H100.txt\", \"9-5-150_30_5th_RAG_data_H100.txt\"]\n",
    "\n",
    "full_file_names = [path + file_name for file_name in file_names]\n",
    "\n",
    "class QuestionAnswerData:\n",
    "    def __init__(self, answer, correct_answer, RAG_data):\n",
    "        self.answer = answer\n",
    "        self.correct_answer = correct_answer\n",
    "        self.is_correct = answer == correct_answer\n",
    "        # RAG data: [{'docId': 'pubmed23n0827_1593', 'BM25_score': 95.4198, 'BM25_ranking': 44, 'SPLADE_score': 54.33488464355469, 'SPLADE_ranking': 13, 'MonoT5_score': 0.9597430488962883, 'MonoT5_ranking': 1}]\n",
    "        self.RAG_data = RAG_data\n",
    "\n",
    "# Get the data\n",
    "\n",
    "# [ [QuestionAnswerData, QuestionAnswerData, ... (1273 data)], [...], [...], [...], [...] ]\n",
    "all_question_answer_data_list = []\n",
    "\n",
    "for i in range(len(full_file_names)):\n",
    "    question_answer_data_list = []\n",
    "    with open(full_file_names[i], \"r\") as f:\n",
    "        count = 0\n",
    "        curr_RAG_data = None\n",
    "        for line in f:\n",
    "            if \"RAG data: \" in line:\n",
    "                curr_RAG_data_str = line[line.find(\"RAG data: \") + len(\"RAG data: \"):].strip()\n",
    "                curr_RAG_data = ast.literal_eval(curr_RAG_data_str)\n",
    "\n",
    "            elif \"question \" in line:\n",
    "                \n",
    "                # Check question number error\n",
    "                count += 1\n",
    "                if \"num:\" in line:\n",
    "                    number = line.strip().split('num:')[-1].split(' ')[0]\n",
    "                else:\n",
    "                    number = line.strip().split('/')[0].split(' ')[-1]\n",
    "                if int(number) != count:\n",
    "                    print(f\"Error: {file_names[i]}:\", number, count)\n",
    "                    break\n",
    "\n",
    "                match = re.search(r\"answer:([^\\s]+)\\s+correct_answer:([^\\s]+)\", line)\n",
    "                if match:\n",
    "                    answer = match.group(1)\n",
    "                    correct_answer = match.group(2)\n",
    "\n",
    "                    question_answer_data = QuestionAnswerData(answer, correct_answer, curr_RAG_data)\n",
    "                    question_answer_data_list.append(question_answer_data)\n",
    "    \n",
    "    all_question_answer_data_list.append(question_answer_data_list)\n",
    "\n",
    "print(len(all_question_answer_data_list[0]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2315c00-de10-4c9f-b8fd-5d6a3cf70784",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (LLM311)",
   "language": "python",
   "name": "llm311"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
